File: src/lib/wow/archive/casc/casc-source-local.ts
/**
 * Local installation CASC source, ported from wow.export (src/js/casc/casc-source-local.js).
 */
import { promises as fsp } from 'fs';
import path from 'path';
import util from 'util';

import { timeEnd, timeLog, write } from '@/lib/wow/log';
import { createProgress, runtimeState } from '@/lib/wow/server/runtime';

import { BufferWrapper } from '../../formats/buffer';
import { constants } from '../../formats/constants';
import { fileExists, readFile } from '../../formats/generics';
import { BLTEReader } from './blte-reader';
import { BuildCache } from './build-cache';
import type { CascKey } from './casc-key';
import { cascKeyFromHex, cascKeyToHex } from './casc-key';
import { CASC } from './casc-source';
import { CASCRemote } from './casc-source-remote';
import { CDNConfigEntries, parseCDNConfig } from './cdn-config';
import { cdnResolver } from './cdn-resolver';
import * as listfile from './listfile';
import { parseVersionConfig } from './version-config';

interface LocalIndexEntry {
  index: number;
  offset: number;
  size: number;
}

export class CASCLocal extends CASC {
  dir: string;

  dataDir: string;

  storageDir: string;

  localIndexes = new Map<string, LocalIndexEntry>();

  cache!: BuildCache;

  remote?: CASCRemote;

  /**
   * Create a new CASC source using a local installation.
   * @param dir Installation path.
   */
  constructor(dir: string) {
    super(false);

    this.dir = dir;
    this.dataDir = path.join(dir, constants.BUILD.DATA_DIR);
    this.storageDir = path.join(this.dataDir, 'data');
  }

  /** Initialize local CASC source. */
  async init(): Promise<void> {
    write('Initializing local CASC installation: %s', this.dir);

    const buildInfo = path.join(this.dir, constants.BUILD.MANIFEST);
    const config = parseVersionConfig(await fsp.readFile(buildInfo, 'utf8'));

    // Filter known products.
    this.builds = config.filter((entry) => constants.PRODUCTS.some((e) => e.product === entry.Product));
  }

  /** Obtain a file by its fileDataID. */
  async getFile(
    fileDataID: number,
    partialDecryption = false,
    suppressLog = false,
    supportFallback = true,
    forceFallback = false,
    contentKey: CascKey | null = null,
  ): Promise<BLTEReader> {
    if (!suppressLog) write('Loading local CASC file %d (%s)', fileDataID, listfile.getByID(fileDataID));

    const encodingKey = contentKey !== null ? this.getEncodingKeyForContentKey(contentKey) : this.getEncodingKey(fileDataID);
    const encodingKeyHex = cascKeyToHex(encodingKey);
    const data = supportFallback
      ? await this.getDataFileWithRemoteFallback(encodingKeyHex, forceFallback)
      : await this.getDataFile(encodingKeyHex);
    return new BLTEReader(data, encodingKeyHex, partialDecryption);
  }

  /**
   * Returns a list of available products in the installation.
   * Format example: "PTR: World of Warcraft 8.3.0.32272"
   */
  getProductList(): string[] {
    const products: string[] = [];
    for (const entry of this.builds) {
      const product = constants.PRODUCTS.find((e) => e.product === entry.Product)!;
      products.push(util.format('%s (%s) %s', product.title, entry.Branch.toUpperCase(), entry.Version));
    }

    return products;
  }

  /** Load the CASC interface with the given build. */
  async load(buildIndex: number): Promise<void> {
    this.resetForLoad();
    this.localIndexes.clear();

    this.build = this.builds[buildIndex];
    write('Loading local CASC build: %o', this.build);

    this.cache = new BuildCache(this.build.BuildKey);
    await this.cache.init();

    this.progress = createProgress(8);
    await this.loadConfigs();
    await this.loadIndexes();
    await this.loadEncoding();
    await this.loadRoot();

    await this.prepareListfile();
    await this.loadListfile(this.build.BuildKey);

    await this.initializeComponents();
    this.isLoaded = true;
  }

  /** Load the BuildConfig from the installation directory. */
  async loadConfigs(): Promise<void> {
    // Load and parse configs from disk with CDN fallback.
    await this.progress.step('Fetching build configurations');

    if (await fileExists('fakebuildconfig')) {
      this.buildConfig = parseCDNConfig(await fsp.readFile('fakebuildconfig', 'utf8'));
      write('WARNING: Using fake build config. No support given for weird stuff happening.');

      // Reconstruct version from the fake config's build name.
      // This is used for e.g. DBD version selection so needs to be correct.
      const splitName = this.buildConfig.buildName.split('patch');
      const buildNumber = splitName[0].replace('WOW-', '');
      const splitPatch = splitName[1].split('_');

      this.build.Version = `${splitPatch[0]}.${buildNumber}`;
    } else {
      this.buildConfig = await this.getConfigFileWithRemoteFallback(this.build.BuildKey);
    }

    this.cdnConfig = await this.getConfigFileWithRemoteFallback(this.build.CDNKey);
  }

  /** Get config from disk with CDN fallback. */
  async getConfigFileWithRemoteFallback(key: string): Promise<CDNConfigEntries> {
    const configPath = this.formatConfigPath(key);
    if (!await fileExists(configPath)) {
      write('Local config file %s does not exist, falling back to CDN...', key);
      if (!this.remote) await this.initializeRemoteCASC();

      const cdnHosts = await cdnResolver.getRankedHosts(runtimeState.selectedCDNRegionTag, this.remote!.serverConfig);
      return this.remote!.getCDNConfig(key, cdnHosts);
    }
    return parseCDNConfig(await fsp.readFile(configPath, 'utf8'));
  }

  /** Load and parse storage indexes from the local installation. */
  async loadIndexes(): Promise<void> {
    timeLog();
    await this.progress.step('Loading indexes');

    let indexCount = 0;

    const entries = await fsp.readdir(this.storageDir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.isFile() && entry.name.endsWith('.idx')) {
        await this.parseIndex(path.join(this.storageDir, entry.name));
        indexCount++;
      }
    }

    timeEnd('Loaded %d entries from %d journal indexes', this.localIndexes.size, indexCount);
  }

  /** Parse a local installation journal index for entries. */
  async parseIndex(file: string): Promise<void> {
    const entries = this.localIndexes;
    const index = await BufferWrapper.readFile(file);

    const headerHashSize = index.readInt32LE();
    index.move(4); // headerHash uint32
    index.move(headerHashSize); // headerHash byte[headerHashSize]

    index.seek((8 + headerHashSize + 0x0F) & 0xFFFFFFF0); // Next 0x10 boundary.

    const dataLength = index.readInt32LE();
    index.move(4);

    const nBlocks = dataLength / 18;
    for (let i = 0; i < nBlocks; i++) {
      const key = index.readHexString(9);
      if (entries.has(key)) {
        index.move(1 + 4 + 4); // idxHigh + idxLow + size
        continue;
      }

      const idxHigh = index.readUInt8();
      const idxLow = index.readInt32BE();

      entries.set(key, {
        index: (idxHigh << 2) | ((idxLow & 0xC0000000) >>> 30),
        offset: idxLow & 0x3FFFFFFF,
        size: index.readInt32LE(),
      });
    }
  }

  /** Load and parse encoding from the local installation. */
  async loadEncoding(): Promise<void> {
    // Parse encoding file.
    timeLog();
    const encKeys = this.buildConfig.encoding.split(' ');

    await this.progress.step('Loading encoding table');
    const encRaw = await this.getDataFileWithRemoteFallback(encKeys[1]);
    this.parseEncodingFile(encRaw, encKeys[1]);
    timeEnd('Parsed encoding table (%d entries)', this.encodingKeys.size);
  }

  /** Load and parse root table from local installation. */
  async loadRoot(): Promise<void> {
    // Get root key from encoding table.
    const rootKey = this.encodingKeys.get(cascKeyFromHex(this.buildConfig.root));
    if (rootKey === undefined) throw new Error('No encoding entry found for root key');

    // Parse root file.
    timeLog();
    await this.progress.step('Loading root file');
    const rootKeyHex = cascKeyToHex(rootKey);
    const root = await this.getDataFileWithRemoteFallback(rootKeyHex);
    const rootEntryCount = this.parseRootFile(root, rootKeyHex);
    timeEnd('Parsed root file (%d entries, %d types)', rootEntryCount, this.rootTypes.length);
  }

  /**
   * Initialize a remote CASC instance to download missing
   * files needed during local initialization.
   */
  async initializeRemoteCASC(): Promise<void> {
    const remote = new CASCRemote(runtimeState.selectedCDNRegionTag);
    await remote.init();

    const buildIndex = remote.builds.findIndex((build) => build.Product === this.build.Product);
    await remote.preload(buildIndex, this.cache);

    this.remote = remote;
  }

  /**
   * Obtain a data file from the local archives.
   * If not stored locally, file will be downloaded from a CDN.
   */
  async getDataFileWithRemoteFallback(key: string, forceFallback = false): Promise<BufferWrapper> {
    try {
      // If forceFallback is true, we have corrupt local data.
      if (forceFallback) throw new Error('Local data is corrupted, forceFallback set.');

      // Attempt 1: Extract from local archives.
      const local = await this.getDataFile(key);

      if (!BLTEReader.check(local)) throw new Error('Local data file is not a valid BLTE');
      validateBLTEData(local, key);

      return local;
    } catch (e) {
      const localError = (e as Error).message;

      // Attempt 2: Load from cache from previous fallback.
      write('Local data file %s could not be used (%s), falling back to cache...', key, localError);
      const cached = await this.cache.getFile(key, constants.CACHE.DIR_DATA);
      if (cached !== null) {
        try {
          validateBLTEData(cached, key);
          return cached;
        } catch (cacheError) {
          write('Cached data file %s is invalid (%s), falling back to CDN...', key, (cacheError as Error).message);
        }
      }

      // Attempt 3: Download from CDN.
      write('Local data file %s not cached or cache invalid, falling back to CDN...', key);
      if (!this.remote) await this.initializeRemoteCASC();

      const remote = this.remote!;
      const archive = remote.archives.get(key);
      let data: BufferWrapper;
      if (archive !== undefined) {
        // Archive exists for key, attempt partial remote download.
        write('Local data file %s has archive, attempt partial download...', key);
        data = await remote.getDataFilePartial(remote.formatCDNKey(archive.key), archive.offset, archive.size);
      } else {
        // No archive for this file, attempt direct download.
        write('Local data file %s has no archive, attempting direct download...', key);
        data = await remote.getDataFile(remote.formatCDNKey(key));
      }

      validateBLTEData(data, key);
      void this.cache.storeFile(key, data, constants.CACHE.DIR_DATA);
      return data;
    }
  }

  /** Obtain a data file from the local archives. */
  async getDataFile(key: string): Promise<BufferWrapper> {
    const entry = this.localIndexes.get(key.substring(0, 18));
    if (!entry) throw new Error(`Requested file does not exist in local data: ${key}`);

    const data = await readFile(this.formatDataPath(entry.index), entry.offset + 0x1E, entry.size - 0x1E);

    let isZeroed = true;
    for (let i = 0, n = data.remainingBytes; i < n; i++) {
      if (data.readUInt8() !== 0x0) {
        isZeroed = false;
        break;
      }
    }

    if (isZeroed) throw new Error(`Requested data file is empty or missing: ${key}`);

    data.seek(0);
    return data;
  }

  /**
   * Format a local path to a data archive.
   * 67 -> <install>/Data/data/data.067
   */
  formatDataPath(id: number): string {
    return path.join(this.dataDir, 'data', `data.${id.toString().padStart(3, '0')}`);
  }

  /**
   * Format a local path to an archive index from the key.
   * 0b45bd2721fd6c86dac2176cbdb7fc5b -> <install>/Data/indices/0b45bd2721fd6c86dac2176cbdb7fc5b.index
   */
  formatIndexPath(key: string): string {
    return path.join(this.dataDir, 'indices', `${key}.index`);
  }

  /**
   * Format a local path to a config file from the key.
   * 0af716e8eca5aeff0a3965d37e934ffa -> <install>/Data/config/0a/f7/0af716e8eca5aeff0a3965d37e934ffa
   */
  formatConfigPath(key: string): string {
    return path.join(this.dataDir, 'config', this.formatCDNKey(key));
  }

  /**
   * Format a CDN key for use in local file reading.
   * Path separators used by this method are platform specific.
   * 49299eae4e3a195953764bb4adb3c91f -> 49\29\49299eae4e3a195953764bb4adb3c91f
   */
  formatCDNKey(key: string): string {
    return path.join(key.substring(0, 2), key.substring(2, 4), key);
  }

  /** Get the current build ID. */
  getBuildName(): string {
    return this.build.Version;
  }

  /** Returns the build configuration key. */
  getBuildKey(): string {
    return this.build.BuildKey;
  }
}

function validateBLTEData(data: BufferWrapper, key: string): void {
  new BLTEReader(data, key);
  data.seek(0);
}

export default CASCLocal;

File: golang/internal/wow/archive/casc/casc_source_local.go
package casc

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/pqhuy98/wow-converter/internal/buffer"
	"github.com/pqhuy98/wow-converter/internal/wow/constants"
	"github.com/pqhuy98/wow-converter/internal/wow/formats"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
	"github.com/pqhuy98/wow-converter/internal/wow/runtime"
)

type localIndexEntry struct {
	index  int
	offset int
	size   int
}

// CASCLocal is a local installation CASC source.
type CASCLocal struct {
	*BaseCASC
	Dir          string
	DataDir      string
	StorageDir   string
	LocalIndexes map[string]localIndexEntry
	Remote       *CASCRemote
}

// NewCASCLocal creates a local CASC source.
func NewCASCLocal(dir string) *CASCLocal {
	base := NewBaseCASC(false)
	dataDir := filepath.Join(dir, constants.Build.DataDir)
	return &CASCLocal{
		BaseCASC:     base,
		Dir:          dir,
		DataDir:      dataDir,
		StorageDir:   filepath.Join(dataDir, "data"),
		LocalIndexes: map[string]localIndexEntry{},
	}
}

// Init initializes the local CASC source.
func (c *CASCLocal) Init() error {
	log.Write("Initializing local CASC installation: %s", c.Dir)
	buildInfo := filepath.Join(c.Dir, constants.Build.Manifest)
	data, err := os.ReadFile(buildInfo)
	if err != nil {
		return err
	}
	config := ParseVersionConfig(string(data))
	for _, entry := range config {
		for _, product := range constants.Products {
			if product.Product == entry["Product"] {
				c.BuildsList = append(c.BuildsList, entry)
				break
			}
		}
	}
	return nil
}

// GetFile obtains a file by fileDataID.
func (c *CASCLocal) GetFile(fileDataID int, partialDecrypt, suppressLog, supportFallback, forceFallback bool, contentKey CascKey) (*BLTEReader, error) {
	if !suppressLog {
		name, _ := GetByID(fileDataID)
		log.Write("Loading local CASC file %d (%s)", fileDataID, name)
	}
	var encodingKey CascKey
	var err error
	if contentKey != "" {
		encodingKey, err = c.GetEncodingKeyForContentKey(contentKey)
	} else {
		encodingKey, err = c.GetEncodingKey(fileDataID)
	}
	if err != nil {
		return nil, err
	}
	encodingKeyHex := CascKeyToHex(encodingKey)
	var data *buffer.Buffer
	if supportFallback {
		data, err = c.GetDataFileWithRemoteFallback(encodingKeyHex, forceFallback)
	} else {
		data, err = c.GetDataFile(encodingKeyHex)
	}
	if err != nil {
		return nil, err
	}
	return NewBLTEReader(data, encodingKeyHex, partialDecrypt)
}

// GetProductList returns available products.
func (c *CASCLocal) GetProductList() []string {
	var products []string
	for _, entry := range c.BuildsList {
		for _, product := range constants.Products {
			if product.Product == entry["Product"] {
				products = append(products, fmt.Sprintf("%s (%s) %s", product.Title, strings.ToUpper(entry["Branch"]), entry["Version"]))
				break
			}
		}
	}
	return products
}

// Load loads the CASC interface with the given build.
func (c *CASCLocal) Load(buildIndex int) error {
	c.ResetForLoad()
	c.LocalIndexes = map[string]localIndexEntry{}
	c.CurrentBuild = c.BuildsList[buildIndex]
	log.Write("Loading local CASC build: %v", c.CurrentBuild)
	c.CacheStore = NewBuildCache(c.CurrentBuild["BuildKey"])
	if err := c.CacheStore.Init(); err != nil {
		return err
	}
	c.ProgressTracker = runtime.CreateProgress(8)
	if err := c.LoadConfigs(); err != nil {
		return err
	}
	if err := c.LoadIndexes(); err != nil {
		return err
	}
	if err := c.LoadEncoding(); err != nil {
		return err
	}
	if err := c.LoadRoot(); err != nil {
		return err
	}
	if err := c.PrepareListfile(); err != nil {
		return err
	}
	if err := c.LoadListfile(); err != nil {
		return err
	}
	if err := c.InitializeComponents(); err != nil {
		return err
	}
	c.Loaded = true
	return nil
}

// LoadConfigs loads build and CDN configs.
func (c *CASCLocal) LoadConfigs() error {
	if err := c.ProgressTracker.Step("Fetching build configurations"); err != nil {
		return err
	}
	if formats.FileExists("fakebuildconfig") {
		data, err := os.ReadFile("fakebuildconfig")
		if err != nil {
			return err
		}
		cfg, err := ParseCDNConfig(string(data))
		if err != nil {
			return err
		}
		c.BuildConfigData = cfg
		log.Write("WARNING: Using fake build config. No support given for weird stuff happening.")
		splitName := strings.Split(c.BuildConfigData["buildName"], "patch")
		buildNumber := strings.TrimPrefix(splitName[0], "WOW-")
		splitPatch := strings.Split(splitName[1], "_")
		c.CurrentBuild["Version"] = splitPatch[0] + "." + buildNumber
	} else {
		cfg, err := c.GetConfigFileWithRemoteFallback(c.CurrentBuild["BuildKey"])
		if err != nil {
			return err
		}
		c.BuildConfigData = cfg
	}
	cfg, err := c.GetConfigFileWithRemoteFallback(c.CurrentBuild["CDNKey"])
	if err != nil {
		return err
	}
	c.CDNConfigData = cfg
	return nil
}

// GetConfigFileWithRemoteFallback gets config from disk with CDN fallback.
func (c *CASCLocal) GetConfigFileWithRemoteFallback(key string) (CDNConfigEntries, error) {
	configPath := c.FormatConfigPath(key)
	if !formats.FileExists(configPath) {
		log.Write("Local config file %s does not exist, falling back to CDN...", key)
		if c.Remote == nil {
			if err := c.InitializeRemoteCASC(); err != nil {
				return nil, err
			}
		}
		cdnHosts, err := DefaultCDNResolver.GetRankedHosts(runtime.RuntimeState.SelectedCDNRegionTag, c.Remote.ServerConfig)
		if err != nil {
			return nil, err
		}
		return c.Remote.GetCDNConfig(key, cdnHosts)
	}
	data, err := os.ReadFile(configPath)
	if err != nil {
		return nil, err
	}
	return ParseCDNConfig(string(data))
}

// LoadIndexes loads and parses storage indexes.
func (c *CASCLocal) LoadIndexes() error {
	log.TimeLog()
	if err := c.ProgressTracker.Step("Loading indexes"); err != nil {
		return err
	}
	indexCount := 0
	entries, err := os.ReadDir(c.StorageDir)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".idx") {
			if err := c.ParseIndex(filepath.Join(c.StorageDir, entry.Name())); err != nil {
				return err
			}
			indexCount++
		}
	}
	log.TimeEnd("Loaded %d entries from %d journal indexes", len(c.LocalIndexes), indexCount)
	return nil
}

// ParseIndex parses a local journal index.
func (c *CASCLocal) ParseIndex(file string) error {
	index, err := buffer.ReadFile(file)
	if err != nil {
		return err
	}
	headerHashSize := int(index.ReadInt32LE().(int64))
	index.Move(4)
	index.Move(headerHashSize)
	index.Seek((8 + headerHashSize + 0x0F) & 0xFFFFFFF0)
	dataLength := int(index.ReadInt32LE().(int64))
	index.Move(4)
	nBlocks := dataLength / 18
	for i := 0; i < nBlocks; i++ {
		key := index.ReadHexString(9)
		if _, exists := c.LocalIndexes[key]; exists {
			index.Move(1 + 4 + 4)
			continue
		}
		idxHigh := int(index.ReadUInt8().(int64))
		idxLow := int(index.ReadInt32BE().(int64))
		c.LocalIndexes[key] = localIndexEntry{
			index:  (idxHigh << 2) | ((idxLow & 0xC0000000) >> 30),
			offset: idxLow & 0x3FFFFFFF,
			size:   int(index.ReadInt32LE().(int64)),
		}
	}
	return nil
}

// LoadEncoding loads and parses the encoding table.
func (c *CASCLocal) LoadEncoding() error {
	log.TimeLog()
	encKeys := strings.Fields(c.BuildConfigData["encoding"])
	if err := c.ProgressTracker.Step("Loading encoding table"); err != nil {
		return err
	}
	encRaw, err := c.GetDataFileWithRemoteFallback(encKeys[1], false)
	if err != nil {
		return err
	}
	if err := c.ParseEncodingFile(encRaw, encKeys[1]); err != nil {
		return err
	}
	log.TimeEnd("Parsed encoding table (%d entries)", c.EncodingTable.len())
	return nil
}

// LoadRoot loads and parses the root table.
func (c *CASCLocal) LoadRoot() error {
	rootKey, ok := c.EncodingTable.lookupEncodingKey(CascKeyFromHex(c.BuildConfigData["root"]))
	if !ok {
		return fmt.Errorf("no encoding entry found for root key")
	}
	log.TimeLog()
	if err := c.ProgressTracker.Step("Loading root file"); err != nil {
		return err
	}
	rootKeyHex := CascKeyToHex(rootKey)
	root, err := c.GetDataFileWithRemoteFallback(rootKeyHex, false)
	if err != nil {
		return err
	}
	rootEntryCount, err := c.ParseRootFile(root, rootKeyHex)
	if err != nil {
		return err
	}
	log.TimeEnd("Parsed root file (%d entries, %d types)", rootEntryCount, len(c.RootTypesList))
	return nil
}

// InitializeRemoteCASC initializes a remote CASC for CDN fallback.
func (c *CASCLocal) InitializeRemoteCASC() error {
	remote := NewCASCRemote(runtime.RuntimeState.SelectedCDNRegionTag)
	if err := remote.Init(); err != nil {
		return err
	}
	buildIndex := -1
	for i, build := range remote.BuildsList {
		if build["Product"] == c.CurrentBuild["Product"] {
			buildIndex = i
			break
		}
	}
	if buildIndex < 0 {
		return fmt.Errorf("remote build not found for product %s", c.CurrentBuild["Product"])
	}
	if err := remote.Preload(buildIndex, c.CacheStore); err != nil {
		return err
	}
	c.Remote = remote
	return nil
}

// GetDataFileWithRemoteFallback obtains a data file with CDN fallback.
func (c *CASCLocal) GetDataFileWithRemoteFallback(key string, forceFallback bool) (*buffer.Buffer, error) {
	tryLocal := func() (*buffer.Buffer, error) {
		if forceFallback {
			return nil, fmt.Errorf("local data is corrupted, forceFallback set")
		}
		local, err := c.GetDataFile(key)
		if err != nil {
			return nil, err
		}
		if !CheckBLTE(local) {
			return nil, fmt.Errorf("local data file is not a valid BLTE")
		}
		return local, nil
	}
	local, err := tryLocal()
	if err == nil {
		return local, nil
	}
	log.Write("Local data file %s does not exist, falling back to cache...", key)
	cached, _ := c.CacheStore.GetFile(key, constants.Cache.DirData)
	if cached != nil {
		return cached, nil
	}
	log.Write("Local data file %s not cached, falling back to CDN...", key)
	if c.Remote == nil {
		if err := c.InitializeRemoteCASC(); err != nil {
			return nil, err
		}
	}
	remote := c.Remote
	archive, ok := remote.Archives[key]
	var data *buffer.Buffer
	if ok {
		log.Write("Local data file %s has archive, attempt partial download...", key)
		data, err = remote.GetDataFilePartial(remote.FormatCDNKey(archive.Key), archive.Offset, archive.Size)
	} else {
		log.Write("Local data file %s has no archive, attempting direct download...", key)
		data, err = remote.GetDataFile(remote.FormatCDNKey(key))
	}
	if err != nil {
		return nil, err
	}
	go c.CacheStore.StoreFile(key, data, constants.Cache.DirData)
	return data, nil
}

// GetDataFile obtains a data file from local archives.
func (c *CASCLocal) GetDataFile(key string) (*buffer.Buffer, error) {
	entry, ok := c.LocalIndexes[key[:18]]
	if !ok {
		return nil, fmt.Errorf("requested file does not exist in local data: %s", key)
	}
	data, err := formats.ReadFile(c.FormatDataPath(entry.index), entry.offset+0x1E, entry.size-0x1E)
	if err != nil {
		return nil, err
	}
	isZeroed := true
	data.Seek(0)
	for i := 0; i < data.RemainingBytes(); i++ {
		if data.ReadUInt8().(int64) != 0 {
			isZeroed = false
			break
		}
	}
	if isZeroed {
		return nil, fmt.Errorf("requested data file is empty or missing: %s", key)
	}
	data.Seek(0)
	return data, nil
}

// FormatDataPath formats a local path to a data archive.
func (c *CASCLocal) FormatDataPath(id int) string {
	return filepath.Join(c.DataDir, "data", fmt.Sprintf("data.%03d", id))
}

// FormatIndexPath formats a local path to an archive index.
func (c *CASCLocal) FormatIndexPath(key string) string {
	return filepath.Join(c.DataDir, "indices", key+".index")
}

// FormatConfigPath formats a local path to a config file.
func (c *CASCLocal) FormatConfigPath(key string) string {
	return filepath.Join(c.DataDir, "config", c.FormatCDNKey(key))
}

// FormatCDNKey formats a CDN key for local file reading.
func (c *CASCLocal) FormatCDNKey(key string) string {
	return filepath.Join(key[:2], key[2:4], key)
}

// GetBuildName returns the current build ID.
func (c *CASCLocal) GetBuildName() string { return c.CurrentBuild["Version"] }

// GetBuildKey returns the build configuration key.
func (c *CASCLocal) GetBuildKey() string { return c.CurrentBuild["BuildKey"] }
